datafusion_ray/core.py

```diff
@@ -0,0 +1,327 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+
+from collections import defaultdict
+from logging import error, debug, info
+import os
+import pyarrow as pa
+import asyncio
+import ray
+import uuid
+import time
+
+from datafusion_ray._datafusion_ray_internal import (
+    RayContext as RayContextInternal,
+    RayDataFrame as RayDataFrameInternal,
+    prettify,
+)
+
+
+def setup_logging():
+    import logging
+
+    logging.addLevelName(5, "TRACE")
+
+    log_level = os.environ.get("DATAFUSION_RAY_LOG_LEVEL", "WARN").upper()
+
+    # this logger gets captured and routed to rust.   See src/lib.rs
+    logging.getLogger("datafusion_ray").setLevel(log_level)
+
+
+setup_logging()
+
+_log_level = os.environ.get("DATAFUSION_RAY_LOG_LEVEL", "ERROR")
+runtime_env = {
+    "worker_process_setup_hook": setup_logging,
+    "env_vars": {"DATAFUSION_RAY_LOG_LEVEL": _log_level, "RAY_worker_niceness": "0"},
+}
+
+
+class RayDataFrame:
+    def __init__(
+        self,
+        ray_internal_df: RayDataFrameInternal,
+        query_id: str,
+        batch_size=8192,
+        isolate_parititions=False,
+        prefetch_buffer_size=0,
+    ):
+        self.df = ray_internal_df
+        self.query_id = query_id
+        self._stages = None
+        self._batches = None
+        self.batch_size = batch_size
+        self.isolate_partitions = isolate_parititions
+        self.prefetch_buffer_size = prefetch_buffer_size
+
+    def stages(self):
```

Comment:
I have two comments here:

- for properties we should use the `@property` decorator, see suggestion above
- we can factor out the function `__init_stages(self)`

i.e. the 

```python
    @property
    def stages(self):
          if not self._stages:
               self._stages = self._init_stages()
          return self.stages
```